[KOGITO-7253] Workflow error handling guide

fjtirado opened the pull request:
**JIRA:**
https://issues.redhat.com/browse/KOGITO-7253

spolti commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -1 +1,77 @@
\ No newline at end of file
+= Understanding workflow error handling
+
+This guide explains how to handle errors that might occur while executing workflow states. Although a non desirable situation, Workflow states, as part of their interaction with external systems, might fail. For these situations, Serverless Workflow specification provides an link:{spec_doc_url}#workflow-error-handling[error handling] mechanism which allows users to deal with these rainy day circu ...
```
I didn't understand the first part correctly.
Handle errors is how to handle errors or handle errors is a method from serverless workflow? (a question that could be asked)

spolti commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -1 +1,77 @@
+
+== Error definition
+
+An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: “wrong parameter”, while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of ...
```
```suggestion
An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: 'wrong parameter', while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of an link:https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html[Exception type] that might be thrown by the state.
```

spolti commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -1 +1,77 @@
+
+An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: “wrong parameter”, while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of ...
+
+The algorithm used by Kogito to determine if an exception should be handled by an error definition is:
```
shouldn't be steps instead algorithm?

fjtirado replied:
Well, al algorithm is another name for a set of rules, isnt it? ;)

fjtirado replied:
ok, changed to steps

spolti commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -1 +1,77 @@
+
+The algorithm used by Kogito to determine if an exception should be handled by an error definition is:
+
+* it checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition. 
```
```suggestion
* It checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition. 
```

spolti commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -1 +1,77 @@
+The algorithm used by Kogito to determine if an exception should be handled by an error definition is:
+
+* it checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition. 
+* if this not the case, it recursively checks if the root cause of the Java Exception is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition.
```
checks if the root cause.... is duplicated from the first bullet, maybe rephrase it?

fjtirado replied:
yes, I normally dislike duplication, but in this case,  rephrasing the second bullet might indicate the procedure is different for the root cause, which is not. 
What about "if this not the case, it recursively performs the previous checks over the the chains of root causes of the Java Exception"?

fjtirado replied:
changed, please take a look @spolti

spolti reviewed (changes requested):
Nice work, just a few comments.

fjtirado commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -1 +1,77 @@
+
+== Error handling example 
+
+Lets illustrate error handling with this link:{kogito_sw_examples_url}/serverless-workflow-error-quarkus[Serverless Workflow Application]. 
```
Too many lets, isnt it? I think the fact the diagram will appear below this paragraph make this lets take a look kind of implicit, wdyt?

spolti replied:
maybe, yes.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -1 +1,77 @@
+     }
+  ]
+----
+* Referring this error definition in the 'checkEven' state to tell the workflow that if this error occurs it should transition to `odd` state rather than the regular transition to `even` state. Since `IllegalArgumentException` is a child of `RuntimeException`, it will be handled by the `errorRef` if thrown by `isEven` function.
```
```suggestion
* Referring to this error definition in the 'checkEven' state to tell the workflow that if this error occurs, it should transition to an `odd` state rather than the regular transition to the `even` state. Since `IllegalArgumentException` is a child of `RuntimeException`, it will be handled by the `errorRef` if thrown by the `isEven` function.
```

ricardozanini reviewed (commented):
Great work! I've added a few minor grammar fixes, but I think Heena will handle this in the future.

I'm not concerned about this guide, but users might find it difficult to map an error within the workflow to a Java Exception without knowing the internals or having Java knowledge. I've opened https://issues.redhat.com/browse/KOGITO-7453 to have a more contextual approach. This is just the beginning, and we can iterate on this later.

hmanwani-rh reviewed (approved):
@fjtirado I've added tweaks to the content and also fixed the image. Please take a look at the applied changes and let me know if we're good to merge this PR. 

Doc preview: https://deploy-preview-31--tangerine-crostata-04a6bf.netlify.app/serverlessworkflow/main/core/understanding-workflow-error-handling.html

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -7,31 +7,43 @@ Understanding workflow error handling
 :java_exception_url: https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html
 
-This guide explains how to handle errors that might occur while executing workflow states. Although a non-desirable situation, Workflow states might fail as part of their interaction with external systems. For these situations, the Serverless Workflow specification provides an link:{spec_doc_url}#workflow-error-handling[error handling] mechanism, allowing users to deal with these rainy day circum ...
+This document describes how you can handle the errors that might occur when you execute your Serverless Workflow states. 
```
Not sure if "when you execute your Serverless Workflow states" passes the right message. It's not the user who executes the states; it's the engine. Too bad if left as "when the Serverless Workflow states are executed"? I know that's passive voice, but it makes more sense to me.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc:
```diff
@@ -7,31 +7,43 @@ Understanding workflow error handling
+image::core/error_handling.png[]
 
-image::core/error_handling.png []
+In the `serverless-workflow-error-quarkus` example application, the workflow accepts an integer number as an input model. When the workflow starts, it invokes the `isEven` Java service, which passes the the defined integer number as a parameter. The link:{kogito_sw_examples_url}serverless-workflow-error-quarkus/src/main/java/org/kie/kogito/examples/EvenService.java[`isEven` Java service] throws a ...
```
```suggestion
In the `serverless-workflow-error-quarkus` example application, the workflow accepts an integer number as an input model. When the workflow starts, it invokes the `isEven` Java service, which passes the defined integer number as a parameter. The link:{kogito_sw_examples_url}serverless-workflow-error-quarkus/src/main/java/org/kie/kogito/examples/EvenService.java[`isEven` Java service] throws an `IllegalArgumentException` exception if the passed integer number is odd. This means that if the integer number is even, the Serverless Workflow transitions to the `even` state, however, if the integer number is odd, the Serverless Workflow transitions to the `odd` state, resulting in a different model output.
```

Small change just to highlight that even and odd, in this context, are state names.

fjtirado replied:
I already changed

ricardozanini reviewed (approved):
@hmanwani-rh @fjtirado I've left a few minor suggestions to the last review.

fjtirado commented:
> @fjtirado I've added tweaks to the content and also fixed the image. Please take a look at the applied changes and let me know if we're good to merge this PR.
> 
@hmanwani-rh There were some things you changed that alter the original meaning, so I have to rewrite again. Next time, please, rather than directly comitting your changes, open a PR over the PR and lets discuss your changes over there.

ricardozanini commented:
I'm merging this PR. Please if we have more divergences, open follow-up patch PR to discuss there.
